### ISSUE TITLE
Should pytest.warns check for issubclass rather than class equality?

### ISSUE THREAD

**Issue by lesteve** [pytest-dev/pytest#2151]
`pytest.warns` check for class equality and not subclass relationship. This snippet for example fails:
```py
import warnings

import pytest


with pytest.warns(Warning):
    warnings.warn('user', UserWarning)
```

This was slightly surprising to me since my expectation was that it would be similar to `pytest.raises` where you can check for a base class:
```py
import pytest


with pytest.raises(Exception):
    raise ValueError()
```

What do you guys think? After all you can always use `with pytest.warns(None) as warninfo` and check `issubclass(warninfo[0].message, a_warning_class)` if you want to check a subclass relationship. At the same time, I don't really how it would hurt to check subclass relationship. 

I guess that is just a matter of changing this [line](https://github.com/pytest-dev/pytest/blob/master/_pytest/recwarn.py#L224) by:
```py
if not any(issubclass(r.category, each) for each in self.expected_warning for r in self):
```
I am willing to open a PR about this if you think it is worth it.

**Comment by lwm** [comment]
[Documentation] reads:

> You can check that code raises a particular warning using `pytest.warns`, 
> which works in a similar manner to raises

It isn't really clear how 'similar' this works since your example shows a clear difference.

Not sure if the functionality should change (seems so, though?) but if it doesn't, then the documentation should be made clearer.

[Documentation]: http://doc.pytest.org/en/latest/recwarn.html#warns

**Comment by lesteve** [comment]
> Not sure if the functionality should change (seems so, though?)

I am kind of leaning this way too. Two additional data points outside of nose.

stdlib warning filters
```py
import warnings

warnings.simplefilter('error', Warning)
warnings.warn('user', UserWarning)  # warning turned into error
```

nose assert_warnings
```py
import nose

with nose.tools.assert_warnings(Warning):
    warnings.warn('user', UserWarning)  # pass because UserWarning is a subclass of Warning
```

**Comment by blueyed** [comment]
@lesteve 
I agree also.
Please go ahead and create a PR.
While it could be seen as a bugfix, I think the features branch is better, since it changes API/behavior.

**Comment by nicoddemus** [comment]
> stdlib warning filters
> nose assert_warnings

Since the other examples do match subclasses, I think it kind makes sense as well (specially with the similarity with `pytest.raises`).


> While it could be seen as a bugfix, I think the features branch is better, since it changes API/behavior.

Agreed!

**Comment by lesteve** [comment]
Closed by #2166.
